### PR TITLE
Rack elevations API

### PR DESCRIPTION
**Pull request by hellerve**
### Fixes: #2248

This PR introduces the rack elevations API. It’s a bit special because it returns SVG results. It replaces the current inline `div`-based solution. I marked it as a draft because there are a few open questions.

I guess an overview of what its capabilities are are in order. Basically, you call `/api/dcim/rack-elevations/<rack-pk>`, and you get back a SVG of that rack’s elevations. There are three URL parameters here:

- `face`: specifies which face to render, `0` for front, `1` for rear. Front is the default.
- `width`: specifies the desired width of the SVG in pixels. 230 is the default.
- `slot_height`: specifies the height of one slot (i.e. one rack unit). I opted for this rather than a total height because the height of racks is usually not as important visually as the height of the devices/components. (I’m not sure about the naming of this param and/or you’d rather like to have total height. Input is appreciated)

The look and feel of the rendered SVG is pretty much identical to that of the DIV, save any bugs or open questions.

### Open Questions

The big one first: what about those tooltips? They’re super useful and I kind of don’t want to lose them, but they’re really not meant to be ported to SVGs. If you twist my arm I can port them, but if we can find a better solution that’d be great!

Then there are a few open questions related to the code climate:
- For the foreground color I’m currently reusing the `fgcolor` utility,w hich is a templatetag. That’s kind of bad style, and I’d opt for extracting that functionality into a helper that is shared by the tempalte tag and the elevation API if desired (it’s desired by me, anyway).
- The stylesheet that we’re using for the rendered SVG is inline, which I think is good (so that the SVG is freestanding). But it’s also inline in the sense of `within the source code`, which is... less ideal. Where should we put it?
- The viewset is fairly readable (to me anyway), but I’m sure it could be improved. Any feedback here is appreciated. A few things that I’m not sure of myself:
  - The GET param validation happens fairly manually. It’s alright I guess, but maybe REST framework has something fancy that we could use?
  - While most of the functions are local to the view set, they are generic helpers that don’t use `self` at all. Should they be factored out somewhere?
  - I’m getting the elevations myself instead of using `rack.get_*_elevation` because I always need both anyway and I was able to simplify the code tremendously by specializing it. At first I tried to reuse the functions, but this leads to more queries and a bunch of glue code to fuse everything. In fact, I was able to remove the `rack.get_*_elevation` functions alltogether, and the `remove_redundant` keyword argument, since they’re now unused. Is the handful of lines of kind-of-but-not-really duplicated code acceptable if we get rid of basically the same amount of code in return?
  - I was liberal in removing legacy code, including stylesheets only used by the elevations and such, and I’m pretty sure nothing broke, but you never know. Is there anything I missed?

All in all, this was equally fun and frustrating, and I hope this is either already useful or can be made useful. I’m looking forward to your review!

Cheers

### CODE REVIEW

**Comment by hellerve** [comment]
The linter was not amused when I forgot to add spaces around my operators. Should be fixed now.

**Comment by lampwins** [comment]
@hellerve this looks really good! I have a couple of high-level comments at this point.

- I think it would be really cool to use this to also expose rack elevations as a dictionary in JSON. Some of your helper methods would actually make this rather trivial.
- With that mind, we should use the `Accept` header to either return an application/json response or the svg. This is more of a pet peeve of mine when REST API's "randomly" change content types without client negotiation.
- If we are only going to render elevations one at a time, would it not make sense to make this a detail action route on the RackViewSet instead of a standalone APIView?
- I understand why you chose to move the helper functions out of the Rack model, but I would actually argue against this because it then limits this low-level API to the REST API only. I.e. I would not be able to use this logic in the nbshell, or my reports, custom scripts, etc, which are officially supported NetBox interfaces.

**Comment by hellerve** [comment]
@lampwins These are all helpful comments! Comment 1 and 2 I’ll try and implement as is.

As for your third comment, I think it makes sense that this is standalone, similarly to how `rack-reservations` is standalone as well. I think they’re either both their own endpoints, or both part of racks. Does that make sense?

As for the fourth comment, I’ll try and put that back into the `Rack` model, though I’m not sure how generally useful it is. Though I understand that even if it’s only useful for one person it’d make sense to have it somewhere where they can use it. I’ll think about naming and whether we can maybe even merge this again with other helper functions, although I’m pretty sure I looked at this when I first worked on it.

**Comment by hellerve** [comment]
@lampwins Something that I didn’t consider is that HTTP Accept headers can’t be manipulated inside tags like `<object>` (and it doesn’t get set by Firefox at least). This means that we need to set `image/svg+xml` as the default value—which would be a little bit annoying in a JSON API—, unless I’m missing an alternative. I vote against loading the SVG in a request made by JS (where we would have that kind of control, but I don’t want to need to enable JS to see the rendered racks).

What do you think?

**Comment by hellerve** [comment]
I drafted something based on `rest_framework` renderers, but sadly this doesn’t solve the more general issue of headers not being set “correctly” by the browser. Once we’ve decided on an alternative I’d be happy to commit and add this to the branch.

For reference, these are the possible solutions that I see:
- Request the SVG from JS.
- Make SVG the default renderer for this view.
- Separate the two endpoints, i.e. making one for JSON and one for SVG.
- Drop the idea of supporting both for now.

**Comment by jeremystretch** [comment]
Great work! I'm going to list my thoughts in no particular order...

* I agree with @lampwins that the SVG rendering code should not be restricted to the API. Ideally it should be accessible directly from the Rack model. (Due to the amount of code required, I'd opt to offload this to a separate class which gets inherited by the `dcim.Rack` model.)

* Per #3569, the rack faces have just recently changed from `0` and `1` to `front` and `rear`, respectively. Just needs a find & replace.

* I would rename `width` and `slot_height` to `unit_width` and `unit_height` for consistency. (NetBox favors the term "unit" for rack openings.)

* The `fgcolor` template tag is a wrapper; you want the underlying utility function at `utilities.util.foreground_color`.

* No problem with using inline style for the SVG but we should move that whole blob of CSS into `dcim/constants.py`.

* I'm not 100% on SVG, but I believe if we embed the SVG code into the HTML directly (as opposed to making an API call for it), we should be able to extend it with tooltips/pop-ups. IIRC that was my original plan. We would just need a toggle on the render method to enable/disable the inclusion of the additional code. (Again, this is just my theory.)

Hoping to share more feedback once I've had the opportunity to try it out. Again, great work!

**Comment by jeremystretch** [comment]
I've re-targeted this PR against a new dedicate branch off of `develop-2.7` (since this is scheduled to be incorporated in the v2.7 release).

I am going to release the v2.7 beta next week, so it would be fantastic to get this wrapped up in the next few days. I'm okay running with it from here if you don't have time @steffann.

**Comment by steffann** [comment]
> I'm okay running with it from here if you don't have time @steffann.

I think you meant @hellerve there :)

**Comment by lampwins** [comment]
@hellerve 

> I think it makes sense that this is standalone, similarly to how rack-reservations is standalone as well.

The difference is rack-reservations is its own model which needs to support creating/modifying/deleting. To me, this elevation feature's implementation is sort of like the napalm integration in the API. The napalm detail route on the device viewset only allows GET operations as a convenience method.

You raise a good point about the accept header method surrounding the actual use case of this endpoint. I would actually be in favor of a query parameter for this. The DRF browsable API view actually does this exact same thing with a query param that allows you to see raw JSON with `?format=json` or the default which renders HTML.

With that in mind I would implement this as a detail action route on the rack model view set:
```
/api/dcim/racks/<id>/elevation/
```
Would return JSON, while passing `?format=svg` would render an SVG:
```
/api/dcim/racks/<id>/elevation/?format=svg
```
That would cover my concerns about client negotiation.

**Comment by hellerve** [comment]
@jeremystretch I might be able to work on this tomorrow (to avoid timezone confusion, that would be European sunday). I’ll keep you posted if I can’t make it, but I should be able to at least do some of the smaller tasks.

**Comment by hellerve** [comment]
So, somehow the retargetting messed up the state of this branch. I’m not quite sure what happened, but I’m going to try and fix it.

File: netbox/dcim/api/serializers.py [hunk 1]
```diff
@@ -157,7 +157,7 @@ class RackUnitSerializer(serializers.Serializer):
     """
     id = serializers.IntegerField(read_only=True)
     name = serializers.CharField(read_only=True)
-    face = serializers.IntegerField(read_only=True)
```
Comment (hellerve) [inline netbox/dcim/api/serializers.py]: It seems like this was forgotten in #3569?

File: netbox/dcim/api/views.py [hunk 2]
```diff
@@ -181,7 +183,7 @@ def units(self, request, pk=None):
         List rack units (by rack)
         """
         rack = get_object_or_404(Rack, pk=pk)
-        face = request.GET.get('face', 0)
+        face = request.GET.get('face', 'front')
```
Comment (hellerve) [inline netbox/dcim/api/views.py]: This one also had to change to fix the tests.

**Comment by hellerve** [comment]
Okay, so I updated this branch with the feedback that you gave me, @jeremystretch. I didn’t manage to also put in the new JSON API or move this to a different class, and I think I won’t have time for that next week, so if you want to finish work on this that’s totally fine with me.

Sorry I wasn’t able to put in all of my changes. I also realized that there is already a way to get rack units in the API and now I’m a little confused as to what the added value of the JSON version of this endpoint is. I’m excited to see what you end up building on top of my janky foundations!

**Comment by lampwins** [comment]
@hellerve great work on this! I am going to merge this into the 2248 branch and clean a few things up. Ultimately my plan is to deprecate the current `units` detail route (in 2.8) in favor of this.

**Pull request merged**